### PR TITLE
libkvmi integration

### PR DESCRIPTION
**Pull request by adlazar**
It builds fine, but I didn't test it yet.
The comments are welcomed.

### CODE REVIEW

File: libvmi/driver/kvm/kvm.c [hunk 1]
```diff
@@ -1436,6 +1623,15 @@ kvm_init_vmi(
     dbprint(VMI_DEBUG_KVM, "--SUCCESS\n");
 #endif
 
+#ifdef HAVE_LIBKVMI
+    dbprint(VMI_DEBUG_KVM, "--Connecting to KVMI...\n");
+    if (!init_kvmi(kvm)) {
+        dbprint(VMI_DEBUG_KVM, "--KVMI failed\n");
+        return VMI_FAILURE;
+    }
+    dbprint(VMI_DEBUG_KVM, "--KVMI connected \o/\n");
```
Comment (Wenzel) [inline libvmi/driver/kvm/kvm.c]: I understand your excitation, i share it, but i think you can remove the emoji \o/ here ;)

File: configure.ac [hunk 2]
```diff
@@ -87,6 +87,13 @@ AC_ARG_ENABLE([kvm],
       [enable_kvm=yes])
 AM_CONDITIONAL([WITH_KVM], [test x"$enable_kvm" = xyes])
 
+AC_ARG_ENABLE([libkvmi],
```
Comment (Wenzel) [inline configure.ac]: I don't think that libkvmi should be enabled or disabled.

To me it **replaces** the current KVM driver of Libvmi, so if the user configured Libvmi to have KVM support, just check that libkvmi is here, or throw an error.

And in the driver there is no need for `HAVE_LIBKVMI` in the same way.

I would like to have @tklengyel opinion too.

**Review by Wenzel — CHANGES_REQUESTED** [review]

**Comment by namidairo** [comment]
This does not appear to be hooked up via the driver interface yet, so it's still gonna call all the old code.

**Comment by adlazar** [comment]
It's more like a demo on how to integrate the kvmi library. :) Maybe I should change the title.
The new functions are called from `kvm_init_vmi`, `kvmi_destroy` and `kvm_get_vcpureg`.

File: configure.ac (hunk 2 above)
Comment (tklengyel) [reply configure.ac]: I would be all for replacing the existing KVM driver with this, but there are features built on top of the existing KVM driver that would need to be removed in that case, like SHM. IMHO that feature was always experimental and not well maintained so removing it should be fine.

**Comment by Wenzel** [comment]
What's the status of this PR ?
What concrete features does it bring ATM ?
Can we merge it already, or are you still polishing the driver ?

**Comment by adlazar** [comment]
@Wenzel, my intention with this PR was to initiate the `libkvmi` integration (init, uninit, domain callbacks ...). It wasn't meant as a complete replacement of the current introspection methods. I see that by not using a proper branch name (i.e. kvmi-begin-libkvmi-integration) and merging @Soft's work it makes this PR look like a replacement. Also, dropping --enable-libkvmi is not an issue, just that I wasn't courageous enough to replace the old introspection methods from this driver :)

I could close this PR if you want to merge it only when is done.

**Comment by Wenzel** [comment]
> @Wenzel, my intention with this PR was to initiate the libkvmi integration (init, uninit, domain callbacks ...). It wasn't meant as a complete replacement of the current introspection methods. I see that by not using a proper branch name (i.e. kvmi-begin-libkvmi-integration) and merging @Soft's work it makes this PR look like a replacement.

I would rather merge what we already have, so we can build tests for libvmi, than continuing adding features on this PR.

This will bring me more visibility, because ATM i don't know what feature you are implementing next @Soft  :)
Let's keep the PR short.
one feature, one PR, and a unit test to validate.

What do you guys think ?

**Review by Wenzel — APPROVED** [review]

**Comment by Wenzel** [comment]
ATM i can't test this PR because `libvmi` won't initialize:
screenshot from 2018-08-06 13-36-31

This check happens [here](https://github.com/libvmi/libvmi/blob/master/libvmi/driver/kvm/kvm.c#L1482), to verify that if the current `QEMU` version is greater than `2.8.0` and we don't have detected `libvmi_request` headers installed by `QEMU` (`/usr/include/qemu/libvmi_request.h`) we exit, because we are supposed to use `libvmi_resquest.h` from `2.8.0` and more for memory access.

screenshot from 2018-08-06 13-43-08

That's why i want to remove all of this code for legacy memory access methods, it's very confusing :)
@tklengyel can give his opinion maybe ?

In the meantime, @Soft created a dedicated branch [kvmi-fixes](https://github.com/Soft/libvmi/tree/kvmi-fixes) which removes this code and allows us to test the driver.

**Comment by Soft** [comment]
Just for the record, there is also another [branch](https://github.com/Soft/libvmi/tree/kvmi-only) that removes all the other KVM introspection methods from the KVM driver leaving only kvmi specific introspection code.

**Comment by adlazar** [comment]
Let's merge https://github.com/Soft/libvmi/tree/kvmi-only. @tklengyel already said (https://github.com/KVM-VMI/libvmi/pull/2/files/f392f4aa6623fb64a5853eef25718427ad5192b3#r197895268) that he is for the replacement of the old KVM driver. Right?

**Comment by Wenzel** [comment]
@adlazar i approve, it's just that @tklengyel's message was confusing, but i think he meant to replace it all.
@Soft can you make a PR to @adlazar 's repo with your `kvmi-only` branch ?
Note: if i remove the QEMU version check, `libkvmi`'s connection fails:
screenshot from 2018-08-06 16-11-54

**Comment by Wenzel** [comment]
This PR is waiting on https://github.com/adlazar/libvmi/pull/8 to be merged.

**Comment by Wenzel** [comment]
Thanks @Soft and @adlazar for merging the `kvmi-only` branch, i'm rebuilding my environment from scratch to test the changes.

**Comment by Wenzel** [comment]
The example is working !
Good job guys :+1: 
confirm

**Comment by namidairo** [comment]
Just skimming through this and noticed that you might have re-introduced the bug where the last part of the memory becomes inaccessible, because of the assumption max memory from libvirt translates to last addressable memory.

I guess that'd break things on a vm using near it's total allocated memory?

**Comment by Wenzel** [comment]
@namidairo do you mean this bug was already fixed in libvmi `master` ?
Can you give me some reference ? (commit, code, ...) because i can't remember it.

**Comment by namidairo** [comment]
https://github.com/libvmi/libvmi/commit/2975d22717b0f51f4e18f9267191aeef691ef123#diff-4ec4791b3c1d65a6821f6146d30de5de

All the QMP code got removed in this PR, so I'm not sure if we'd put it back just for this or just expand on the qemu patch since kvmi has one already.

**Pull request merged**